src/main/java/io/cdap/plugin/servicenow/sink/service/ServiceNowSinkAPIRequestImpl.java

```diff
@@ -0,0 +1,259 @@
+import com.google.gson.JsonParser;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.plugin.servicenow.apiclient.ServiceNowTableAPIClientImpl;
+import io.cdap.plugin.servicenow.apiclient.ServiceNowTableAPIRequestBuilder;
+import io.cdap.plugin.servicenow.restapi.RestAPIResponse;
+import io.cdap.plugin.servicenow.sink.ServiceNowSinkConfig;
+import io.cdap.plugin.servicenow.sink.model.RestRequest;
+import io.cdap.plugin.servicenow.sink.model.ServiceNowBatchRequest;
+import io.cdap.plugin.servicenow.util.ServiceNowConstants;
+
+import org.apache.http.Header;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.oltu.oauth2.common.exception.OAuthProblemException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import javax.ws.rs.core.MediaType;
+
+/**
+ * Implementation class for ServiceNow Batch Rest API.
+ */
+public class ServiceNowSinkAPIRequestImpl {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowSinkAPIRequestImpl.class);
+
+  private static Integer counter = 1;
+  private final ServiceNowSinkConfig config;
+  private final ServiceNowTableAPIClientImpl restApi;
+  private final Gson gson = new Gson();
+  private final JsonParser jsonParser = new JsonParser();
+
+  public ServiceNowSinkAPIRequestImpl(ServiceNowSinkConfig conf) {
+    this.config = conf;
+    restApi = new ServiceNowTableAPIClientImpl(config);
+  }
+
+  public RestRequest getRestRequest(JsonObject jsonObject) {
+    JsonElement jsonElement = jsonObject.get(ServiceNowConstants.SYS_ID);
+    String sysId = null;
+    // sys_id is mandatory for update operation
+    if (jsonElement == null && config.getOperation().equals(ServiceNowConstants.UPDATE_OPERATION)) {
+      throw new IllegalStateException("No sys_id found in the record to be updated");
+      // sys_id need to be removed for insert operation
+    } else if (jsonElement != null && config.getOperation().equals(ServiceNowConstants.INSERT_OPERATION)) {
+      jsonObject.remove(ServiceNowConstants.SYS_ID);
+      // this condition will run when jsonElement is not null and operation is update
+    } else if (jsonElement != null) {
+      sysId = jsonElement.getAsString();
+    }
+    String data = jsonObject.toString();
+    String encodedData = Base64.getEncoder().encodeToString(data.getBytes(StandardCharsets.UTF_8));
+
+    RestRequest restRequest = new RestRequest();
+    restRequest.setUrl(getUrl(sysId));
+    restRequest.setId(counter.toString());
```

Comment:
this is not thread safe - can end up with a duplicate id if this is called concurrently from multiple threads. 
If you just need a unique id, then you can use a uuid.

Reply:
Making use of `AtomicInteger` now, for Request ID. bc67f6e1b45ebd37026d24f7b64a2d8d43fbbc88